rs/index/src/segment/mutable_segment.rs

```diff
@@ -38,12 +41,16 @@ impl MutableSegment {
             return Err(anyhow::anyhow!("Cannot insert into a finalized segment"));
         }
 
+        if let Some(invalidated_id_set) = self.invalidated_ids.get_mut(&0) {
+            // It's ok if the doc_id hasn't been invalidated before
+            invalidated_id_set.remove(&doc_id);
+        }
         self.insert_for_user(0, doc_id, data, 0)
     }
 
     /// Insert a document for a user
     pub fn insert_for_user(
-        &self,
+        &mut self,
```

Comment:
let's not use `mut` here - try to play with interior mutability

Reply:
Nice, thanks for the suggestion